[inductor] Lower masked_scatter on CUDA

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108804
* __->__ #108803

This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

peterbell10 committed:
[inductor] Lower masked_scatter on CUDA

This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Lower masked_scatter on CUDA"


This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Lower masked_scatter on CUDA"


This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Lower masked_scatter on CUDA"


This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Lower masked_scatter on CUDA"


This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -2740,6 +2740,42 @@ def output_indexer(index):
+            i = ops.indirect_indexing(source_idx_val, source_numel)
+            return source_loader([i])
+
+        source_val = ops.masked(mask_val, load_source_val, 0)
+        return ops.where(mask_val, source_val, self_val)
```
In other places in the lowerings we do a `where` after a `masked`, but I think it should be a tad more efficient to do an `ops.add`.

peterbell10 replied:
`ops.masked` under the hood returns `tl.where(mask, val, 0)`, so by doing another `tl.where` with the same mask value, I expect this to be optimized into a single `tl.where` by the triton compiler.

lezcano replied:
Could you confirm that this is the case? If it is, we should probably change the lowering that does a masked into an add.

peterbell10 replied:
I'm slightly surprised how far this makes it through the optimization pipeline before getting removed. It goes all the way to the llvm IR:
```mlir
    %74 = llvm.select %44, %73, %8 : i1, f32 loc(#loc27)
    %75 = llvm.select %44, %74, %37 : i1, f32 loc(#loc28)
```

But it is optimized away by the time we get to the ptx.
```ptx
        @%p11 ld.global.b32 { %r9 }, [ %rd23 + 0 ];
        @!%p11 mov.u32 %r9, %r10;
        .loc    1 28 32
        selp.b32        %r11, %r9, %r3, %p6;
        .loc    1 29 25
        shl.b64         %rd26, %rd1, 2;
        add.s64         %rd24, %rd4, %rd26;
        .loc    1 29 36
        @%p13 st.global.b32 [ %rd24 + 0 ], { %r11 };
```

Here you can see only one select op (`selp.b32`) between loading `source_val` (`%r9`) and storing the final result (`%r11`).

> we should probably change the lowering that does a masked into an add.

All I could find was https://github.com/pytorch/pytorch/blob/5531a23b204b4daa2c0bb3c52610e9a0ba79dacf/torch/_inductor/lowering.py#L3199-L3200

which seems to be accumulating a gradient so an add really is what we want. lmk if you had something in mind.

lezcano replied:
yeah, it's true that that `add` is necessary. Thank you for looking into this.

lezcano replied:
I hope the CPU path can also optimise this out (no need to check). It would also be possible to make `masked`  accept a tensor as a second argument as
```python
ops.masked(mask_val, load_source_val, self_val)
```
but yeah, no need.

lezcano commented on torch/_inductor/decomposition.py:
```diff
@@ -550,3 +550,12 @@ def select_decomp_table():
+
+@register_decomposition(aten.masked_scatter)
+def masked_scatter(self, mask, source):
+    if self.device.type == "cuda":
```
Could you leave a comment here as well as to why do we just decompose CUDA but not CPU.

peterbell10 committed:
Update on "[inductor] Lower masked_scatter on CUDA"


This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 committed:
Lint on "[inductor] Lower masked_scatter on CUDA"


This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

lezcano commented:
@pytorchbot revert -m "makes test_comprehensive_masked_scatter_cuda_int64 flaky" -c nosignal

lezcano commented:
See https://github.com/pytorch/pytorch/actions/runs/6164115376/job/16729869079

lezcano commented:
There's also `PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_torch.py -k test_masked_scatter_inplace_noncontiguous_cuda` failing

peterbell10 commented:
@pytorchbot merge -r

peterbell10 commented:
@pytorchbot revert -c weird -m "Bot merged after aborted rebase"

peterbell10 committed:
Update on "[inductor] Lower masked_scatter on CUDA"


This decomposes masked_scatter into `aten.cumsum` and a single pointwise kernel,
which is similar to what is done in eager. I only do this for CUDA because on CPU
it isn't split into two passes like this so would cause a slowdown.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
